File: python/tempmail_wrapper/providers/one_sec_email.py
"""1secemail.com provider implementation (web scraping with CSRF)."""

from __future__ import annotations

import random
import re
import string
from datetime import datetime
from html import unescape

from ..base import TempMailProvider
from ..exceptions import NotFoundError, TempMailError
from ..http_client import HttpClient
from ..models import Message, MessageDetail

DOMAINS = [
    "qzueos.com", "gaziw.com", "emailgenerator.xyz",
]


class OneSecEmailProvider(TempMailProvider):
    """Provider for 1secemail.com — CSRF-protected web scraping."""

    BASE = "https://www.1secemail.com"

    def __init__(
        self,
        proxies: list[str] | None = None,
        random_ua: bool = True,
        **kwargs,
    ) -> None:
        self._http = HttpClient(proxies=proxies, random_ua=random_ua, use_cookies=True)
        self._csrf: str | None = None
        self._email: str | None = None

    @property
    def name(self) -> str:
        return "1secemail"

    def _fetch_csrf(self) -> str:
        resp = self._http.get(f"{self.BASE}/", timeout=30)
        if resp.status_code != 200:
            raise TempMailError(f"Failed to load page: {resp.status_code}", self.name)
        m = re.search(r'<meta name="csrf-token" content="([^"]+)">', resp.text)
        if not m:
            raise TempMailError("CSRF token not found", self.name)
        return m.group(1)

    def _ensure_csrf(self) -> None:
        if self._csrf:
            return
        self._csrf = self._fetch_csrf()

    def _post(self, url: str, data: dict) -> dict:
        self._ensure_csrf()
        resp = self._http.post(
            url,
            json={"_token": self._csrf, **data},
            headers={
                "X-CSRF-TOKEN": self._csrf,
                "x-xsrf-token": self._csrf,
                "Referer": f"{self.BASE}/",
            },
            timeout=30,
        )
        if resp.status_code != 200:
            raise TempMailError(f"POST failed: {resp.status_code}", self.name)
        try:
            return resp.json()
        except Exception:
            return {}

    def generate_email(self) -> str:
        self._ensure_csrf()
        name = "".join(random.choices(string.ascii_lowercase + string.digits, k=10))
        domain = random.choice(DOMAINS)
        self._post(f"{self.BASE}/change", {"name": name, "domain": domain})
        self._email = f"{name}@{domain}"
        return self._email

    def get_inbox(self, email: str) -> list[Message]:
        self._ensure_csrf()
        data = self._post(f"{self.BASE}/get_messages", {})
        if not isinstance(data, list):
            # API might return { "messages": [...] } or { "data": [...] }
            if isinstance(data, dict):
                data = data.get("messages", data.get("data", []))
            else:
                return []
        msgs = []
        for m in data:
            date_str = m.get("receivedAt", "")
            try:
                date = datetime.strptime(date_str, "%Y-%m-%d %H:%M:%S")
            except (ValueError, TypeError):
                date = datetime.now()
            msgs.append(Message(
                id=m.get("id", ""),
                sender=m.get("from_email", m.get("from", "unknown")),
                subject=m.get("subject", "(no subject)"),
                date=date,
            ))
        return msgs

    def read_message(self, message_id: str) -> MessageDetail:
        self._ensure_csrf()
        resp = self._http.get(
            f"{self.BASE}/view/{message_id}",
            headers={"X-CSRF-TOKEN": self._csrf, "Referer": f"{self.BASE}/"},
            timeout=30,
        )
        if resp.status_code != 200:
            raise NotFoundError(f"Message {message_id} not found", self.name)
        html = resp.text
        text = re.sub(r"<[^>]+>", "", unescape(html))
        text = re.sub(r"\s+", " ", text).strip()
        sender_match = re.search(r"From:\s*([^<\n]+)", html)
        subject_match = re.search(r"Subject:\s*([^<\n]+)", html)
        try:
            date_match = re.search(r"(\d{4}-\d{2}-\d{2}\s+\d{2}:\d{2}:\d{2})", html)
            date = datetime.strptime(date_match.group(1), "%Y-%m-%d %H:%M:%S") if date_match else datetime.now()
        except ValueError:
            date = datetime.now()
        return MessageDetail(
            message=Message(
                id=message_id,
                sender=sender_match.group(1).strip() if sender_match else "unknown",
                subject=subject_match.group(1).strip() if subject_match else "(no subject)",
                date=date,
            ),
            body_text=text,
            body_html=html,
        )

    def delete_email(self, email: str) -> bool:
        return True

File: go/providers/onesecemail.go
package providers

import (
	"encoding/json"
	"fmt"
	"math/rand"
	"net/http"
	"regexp"
	"strings"
	"time"
)

var oseDomains = []string{
	"qzueos.com", "gaziw.com", "emailgenerator.xyz",
}

// OneSecEmail is the 1secemail.com provider (CSRF-protected web scraping).
type OneSecEmail struct {
	client *HttpClient
	csrf   string
	email  string
}

// NewOneSecEmail creates a new OneSecEmail provider.
func NewOneSecEmail(config map[string]string) *OneSecEmail {
	proxies := ParseProxies(config["proxies"])
	randomUA := config["randomUa"] == "true"
	return &OneSecEmail{
		client: NewHttpClient(proxies, randomUA, true),
	}
}

func (o *OneSecEmail) fetchCSRF() error {
	resp, err := o.client.Get("https://www.1secemail.com/")
	if err != nil {
		return err
	}
	body := string(resp)
	m := regexp.MustCompile(`<meta name="csrf-token" content="([^"]+)">`).FindStringSubmatch(body)
	if len(m) < 2 {
		return &TempMailError{Provider: "1secemail", Message: "CSRF token not found"}
	}
	o.csrf = m[1]
	return nil
}

func (o *OneSecEmail) ensureCSRF() error {
	if o.csrf != "" {
		return nil
	}
	return o.fetchCSRF()
}

func (o *OneSecEmail) postForm(urlPath string, data map[string]string) ([]byte, error) {
	if err := o.ensureCSRF(); err != nil {
		return nil, err
	}
	form := make(map[string]string)
	form["_token"] = o.csrf
	for k, v := range data {
		form[k] = v
	}
	body, _ := json.Marshal(form)
	req, _ := http.NewRequest(http.MethodPost, urlPath, strings.NewReader(string(body)))
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-CSRF-TOKEN", o.csrf)
	req.Header.Set("x-xsrf-token", o.csrf)
	req.Header.Set("Referer", "https://www.1secemail.com/")
	return o.client.RawDo(req, body)
}

func randomOSEName() string {
	const chars = "abcdefghijklmnopqrstuvwxyz0123456789"
	b := make([]byte, 10)
	for i := range b {
		b[i] = chars[rand.Intn(len(chars))]
	}
	return string(b)
}

func (o *OneSecEmail) GenerateEmail() (string, error) {
	if err := o.ensureCSRF(); err != nil {
		return "", err
	}
	name := randomOSEName()
	domain := oseDomains[rand.Intn(len(oseDomains))]
	_, err := o.postForm("https://www.1secemail.com/change", map[string]string{"name": name, "domain": domain})
	if err != nil {
		return "", err
	}
	o.email = name + "@" + domain
	return o.email, nil
}

func (o *OneSecEmail) GetInbox(email string) ([]Message, error) {
	resp, err := o.postForm("https://www.1secemail.com/get_messages", map[string]string{})
	if err != nil {
		return nil, err
	}

	var rawItems []json.RawMessage
	// Try as array first
	if err := json.Unmarshal(resp, &rawItems); err != nil {
		// Try as object, look for array fields
		var obj map[string]json.RawMessage
		if err2 := json.Unmarshal(resp, &obj); err2 != nil {
			return nil, &TempMailError{Provider: "1secemail", Message: "parse error: " + err.Error()}
		}
		// Common key names for message arrays
		for _, key := range []string{"messages", "data", "emails", "inbox"} {
			if v, ok := obj[key]; ok {
				if err3 := json.Unmarshal(v, &rawItems); err3 == nil {
					break
				}
			}
		}
		if len(rawItems) == 0 {
			return []Message{}, nil
		}
	}

	var items []struct {
		ID         string `json:"id"`
		From       string `json:"from"`
		FromEmail  string `json:"from_email"`
		Subject    string `json:"subject"`
		ReceivedAt string `json:"receivedAt"`
	}
	for _, ri := range rawItems {
		var it struct {
			ID         string `json:"id"`
			From       string `json:"from"`
			FromEmail  string `json:"from_email"`
			Subject    string `json:"subject"`
			ReceivedAt string `json:"receivedAt"`
		}
		if err := json.Unmarshal(ri, &it); err != nil {
			continue
		}
		items = append(items, it)
	}

	msgs := make([]Message, 0, len(items))
	for _, it := range items {
		sender := it.FromEmail
		if sender == "" {
			sender = it.From
		}
		if sender == "" {
			sender = "unknown"
		}
		dt, _ := parseOSEDate(it.ReceivedAt)
		msgs = append(msgs, Message{
			ID:      it.ID,
			Sender:  sender,
			Subject: firstOfOse(it.Subject, "(no subject)"),
			Date:    dt,
		})
	}
	return msgs, nil
}

func (o *OneSecEmail) ReadMessage(messageID string) (*MessageDetail, error) {
	if err := o.ensureCSRF(); err != nil {
		return nil, err
	}
	resp, err := o.client.Get(fmt.Sprintf("https://www.1secemail.com/view/%s", messageID))
	if err != nil {
		return nil, err
	}
	html := string(resp)
	text := stripOSEHTML(html)

	var sender, subject string
	date := time.Now()
	if m := regexp.MustCompile(`From:\s*([^<\n]+)`).FindStringSubmatch(html); len(m) > 1 {
		sender = strings.TrimSpace(m[1])
	}
	if sender == "" {
		sender = "unknown"
	}
	if m := regexp.MustCompile(`Subject:\s*([^<\n]+)`).FindStringSubmatch(html); len(m) > 1 {
		subject = strings.TrimSpace(m[1])
	}
	if subject == "" {
		subject = "(no subject)"
	}
	if m := regexp.MustCompile(`(\d{4}-\d{2}-\d{2}\s+\d{2}:\d{2}:\d{2})`).FindStringSubmatch(html); len(m) > 1 {
		date, _ = parseOSEDate(m[1])
	}
	return &MessageDetail{
		Message: Message{
			ID:      messageID,
			Sender:  sender,
			Subject: subject,
			Date:    date,
		},
		BodyText: text,
		BodyHTML: html,
	}, nil
}

func (o *OneSecEmail) DeleteEmail(email string) (bool, error) {
	return true, nil
}

func (o *OneSecEmail) WaitForEmail(email string, timeout time.Duration, interval time.Duration) (*Message, error) {
	return waitForEmail(o, email, timeout, interval)
}

func parseOSEDate(s string) (time.Time, error) {
	return time.Parse("2006-01-02 15:04:05", s)
}

func stripOSEHTML(s string) string {
	re := regexp.MustCompile(`<[^>]+>`)
	s = re.ReplaceAllString(s, "")
	s = strings.ReplaceAll(s, "\n", " ")
	s = regexp.MustCompile(`\s+`).ReplaceAllString(s, " ")
	return strings.TrimSpace(s)
}

func firstOfOse(vals ...string) string {
	for _, v := range vals {
		if v != "" {
			return v
		}
	}
	return ""
}
